### ISSUE TITLE
Animated lights appear to be blended incorrectly with the static lightmap, resulting in odd colours

### ISSUE THREAD

**Issue by andyp123** [sezero/quakespasm#47]
This image shows a scene with two lights in Ironwail (it looks the same in vkQuake, and similar in the KEX remaster engine). One light is not animated (style 0) and the other is set to pulse slowly (style 2). The result is that the animated light appears to be blended correctly with the static light, leading to a very bright light near the lava surface.
![ironwail0008](https://user-images.githubusercontent.com/1097744/215337222-81edfde4-6aa6-497e-a65b-82ff60a390a4.png)

However, in Quakespasm, the blend results in a strange green colour.
![quakespasm](https://user-images.githubusercontent.com/1097744/215337197-3d978726-6e35-486f-ae91-41dea002a1cf.jpg)

This seems to be most noticable with orange lights (lava) and grey textures. I didn't notice the problem nearly as much if I changed the texture to something brighter or more saturated. I could reduce the green tint in Quakespasm by lowering the brightness of the lights significantly, but it negatively affected the visuals in other engines.

There was some discussion about the issue in the Quake Mapping discord, and Pyro posted [this relevant link](https://kneedeepinthedoomed.wordpress.com/2010/11/07/orange-light-green-walls-far-out-experiences-with-colored-light/).

You can get the test map [here](https://drive.google.com/file/d/1HEI-L5ONgxVd9uKOdPNr6rXa9otcxHT7/view?usp=sharing).

**Comment by sezero** [comment]
@andrei-drexler, @ericwa ?

**Comment by ericwa** [comment]
Haven't confirmed, but I think this is from clamping to bytes at: https://github.com/sezero/quakespasm/blob/master/Quake/r_brush.c#L1241 and  to support this, we'd need to instead use a format like RBGA_10_10_10_2 and clamp at 1023 instead of 255

**Comment by ericwa** [comment]
Actually, there's more to it than that. Test map with 4 stacked red spotlights, 3 of them are switchable. The ability of Ironwail to add the 4 lightstyles together without clamping at 2.0 is visible, but note that vkQuake renders it like Quakespasm:

![image](https://user-images.githubusercontent.com/239161/215354806-f263c023-2a4e-46e3-a3e0-bd0d3aa98ddc.png)
[lightmap_blending.zip](https://github.com/sezero/quakespasm/files/10530815/lightmap_blending.zip)

However with andyp's testmap, vkQuake doesn't turn green:

![image](https://user-images.githubusercontent.com/239161/215355024-7d7c23b8-4c10-40d9-a66e-389e7cdb689d.png)

Not sure what's going on..

**Comment by sezero** [comment]
> Actually, there's more to it than that. Test map with 4 stacked red spotlights, 3 of them are switchable. The ability of Ironwail to add the 4 lightstyles together without clamping at 2.0 is visible, but note that vkQuake renders it like Quakespasm:
> 
> However with andyp's testmap, vkQuake doesn't turn green:
> 
> Not sure what's going on..

@Novum ?

**Comment by Flecked42** [comment]
The plot thickens. I removed the styles from the lights and recompiled the map. [lightbug2.zip](https://github.com/sezero/quakespasm/files/10540447/lightbug2.zip)


This is what happens with `gl_overbright` on and off.

![spasm0000](https://user-images.githubusercontent.com/90434262/215608026-b0cf1808-adee-4ffd-b097-57e156e5e712.png)
![spasm0001](https://user-images.githubusercontent.com/90434262/215608086-c1caebe8-a519-4ec5-9a89-5a76da09aecf.png)

**Comment by andrei-drexler** [comment]
> Actually, there's more to it than that. Test map with 4 stacked red spotlights, 3 of them are switchable. The ability of Ironwail to add the 4 lightstyles together without clamping at 2.0 is visible, but note that vkQuake renders it like Quakespasm:

I think `r_gpulightmapupdate` may have been disabled in vkQuake for the lightmap_blending screenshots (since the CPU code path doesn't seem to have the same extended range as the GPU path at the moment):
| r_gpulightmapupdate 0 | r_gpulightmapupdate 0 |
|---|---|
|![vkquake0024](https://user-images.githubusercontent.com/5121829/216821745-96130f20-ad1a-442b-a0cd-6ef844e02246.png)|![vkquake0023](https://user-images.githubusercontent.com/5121829/216821763-01c0001c-b4ed-43af-9bb2-020b6fe2702c.png)|

**Comment by ericwa** [comment]
Good catch! I didn't change `r_gpulightmapupdate` by hand, but I was rotating between those 4 engines (+ also ran an old vkQuake version) when collecting those earlier screenshots, and it must have been toggled inadvertently.

Testing `lightbug.bsp` from the first post in the issue again, with `r_gpulightmapupdate 1`:

![vkquake0006](https://user-images.githubusercontent.com/239161/216839114-199644ec-2564-43fb-9049-b9ea011a07d9.png)

and `r_gpulightmapupdate 0`:

![vkquake0007](https://user-images.githubusercontent.com/239161/216839115-8ef33ed5-83f9-4e60-88ca-78c5a1c5c5c5.png)

So, it looks like the green lava glow is a symptom of limited range lightmaps and my original idea should work for QS https://github.com/sezero/quakespasm/issues/47#issuecomment-1407741722

**Comment by Novum** [comment]
While at it, 10 bit for the color buffer also removes some banding problems in dark/foggy areas.

**Comment by Flecked42** [comment]
Joe applied @andrey-budko fix to JoeQuake https://github.com/j0zzz/JoeQuake/commit/a6192c07577b6b40348f0c71c6358ada36e134df and it seems to work good expect there's some lighting artifacts with AMD cards.

Here's how it looks for me with AMD card.
https://github.com/sezero/quakespasm/assets/90434262/55f0fac3-ac15-469b-a5a3-be19bf7d3ac5

Here's how it looks for Joe with NVIDIA card.
https://github.com/sezero/quakespasm/assets/90434262/69525e09-3828-4977-932d-0ae361dd8d20

Any ideas?

**Comment by andrey-budko** [comment]
@Flecked42 did you try patched quakespasm?

This is mingw build if you want
https://prboom-plus.sourceforge.io/quakespasm-issue-47.zip

<details>
<summary>master vs patched</summary>

 ![image](https://github.com/sezero/quakespasm/assets/6490190/8d671225-a3cf-41be-ad24-7e138010d76f)want inside the   
</details>

**Comment by Flecked42** [comment]
@andrey-budko I get the same artifacts with the patched QS

**Comment by Flecked42** [comment]
I tried `GL_UNSIGNED_INT_2_10_10_10_REV` instead of `GL_UNSIGNED_INT_10_10_10_2` and it seemed to fix the artifacts but the colors are messed up

![joequake000](https://github.com/sezero/quakespasm/assets/90434262/8664b994-b797-44e8-b139-6145d43fcdf0)

**Comment by ericwa** [comment]
Documentation on this stuff is lacking, but https://stackoverflow.com/questions/17726676/what-does-rev-suffix-mean-in-opengl suggests `GL_UNSIGNED_INT_2_10_10_10_REV` + `GL_RGBA` (that is gl_lightmap_format in joequake) should be packed like:

`*(unsigned int*)dest = (3 << 30) | (b << 20) | (g << 10) | r;`

@Flecked42 Try that on line 439 when setting `GL_UNSIGNED_INT_2_10_10_10_REV` maybe?

**Comment by Flecked42** [comment]
@ericwa Added `*(unsigned int*)dest = (3 << 30) | (b << 20) | (g << 10) | r;` in and the colors are right now, but the artifacts came back.

**Comment by andrey-budko** [comment]
@Flecked42, how about this change?
```patch
-	glTexImage2D (GL_TEXTURE_2D, 0, lightmap_bytes, glt->width, glt->height, 0, gl_lightmap_format, GL_UNSIGNED_BYTE, data);
+	glTexImage2D (GL_TEXTURE_2D, 0, GL_RGB10_A2, glt->width, glt->height, 0, gl_lightmap_format, GL_UNSIGNED_INT_10_10_10_2, data);
```

i mean changing `lightmap_bytes` with `GL_RGB10_A2` in `gl_texmgr.c` / `TexMgr_LoadLightmap` / `glTexImage2D `

**Comment by andrey-budko** [comment]
Got the same lighting artifacts on Intel HD Graphics 4600. Replacing  `internalformat` for `glTexImage2D` with `GL_RGB10_A2` fixed the issue for me.

I've created a branch in case anyone wants to experiment
https://github.com/andrey-budko/ironwail/commit/4e216c290dca379b013c06d104ff132297a987ab

**Comment by j0zzz** [comment]
Thank you @andrey-budko ❤️

**Comment by sezero** [comment]
@ericwa: Is suggested patch https://github.com/andrey-budko/ironwail/commit/4e216c290dca379b013c06d104ff132297a987ab good? (It requires GL_EXT_packed_pixels - I guess checking for it would be not necessary these days.)

**Comment by sezero** [comment]
With @andrey-budko patch and using `-noglslalias`, I get a lot darker scene than without.

**Comment by Flecked42** [comment]
JoeQuake got abit darker as well. Maybe a side effect of the patch?

**Comment by Flecked42** [comment]
JoeQuake got abit darker as well.

Maybe increasing this value abit would do it https://github.com/andrey-budko/ironwail/blob/4e216c290dca379b013c06d104ff132297a987ab/Quake/r_world.c#L856C1-L856C1

**Comment by j0zzz** [comment]
@andrey-budko adapted the rendering part for the GLSL script only. When -noglslalias or -nomtex command line parameters are used, lightmaps are rendered differently and that part of the code was not changed. Check my commit for JoeQuake to see how I fixed them: https://github.com/j0zzz/JoeQuake/commit/a6192c07577b6b40348f0c71c6358ada36e134df

When the lightmaps are drawn via multitexture, you need to replace
`glTexEnvf (GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_MODULATE);`
with
`glTexEnvf (GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_COMBINE);
glTexEnvf (GL_TEXTURE_ENV, GL_COMBINE_RGB, GL_MODULATE);
glTexEnvf (GL_TEXTURE_ENV, GL_RGB_SCALE, 4);`
See the part under `R_DrawTextureChains_Multitexture` in my commit.

And when lightmaps are drawn as a separate pass, then you need to **add** the above mentioned 3 lines of `glTexEnvf` before rendering the lightmaps. See the my changes under `R_BlendLightmaps` in my commit.
@sezero

**Comment by sezero** [comment]
As far as I can see, the joequake and quakespasm code paths don't fully look alike and I don't want to get confused more than I already am. The GL_COMBINE and GL_RGB_SCALE changes definitely require GL_EXT_texture_env_combine so that adds an extra complexity of checks I guess. (The GL_UNSIGNED_INT_10_10_10_2 changes already require GL_EXT_packed_pixels which I _do_ have a patch for checking it before use, but not for the non-glsl path..)

In short: Unless someone gives me a _well-tested_ patch, this issue can wait (and the issue _is_ minor already.)

**Comment by sezero** [comment]
> As far as I can see, the joequake and quakespasm code paths don't fully look alike and I don't want to get confused more than I already am. The GL_COMBINE and GL_RGB_SCALE changes definitely require GL_EXT_texture_env_combine so that adds an extra complexity of checks I guess. (The GL_UNSIGNED_INT_10_10_10_2 changes already require GL_EXT_packed_pixels which I _do_ have a patch for checking it before use, but not for the non-glsl path..)
> 
> In short: Unless someone gives me a _well-tested_ patch, this issue can wait (and the issue _is_ minor already.)

BTW, here is the patch I tested against quakespasm git master with packed_pixel extension checks in place:
[47-x.patch.txt](https://github.com/sezero/quakespasm/files/12373583/47-x.patch.txt)

One lazy solution I can think of for the non-GLSL path is enabling `gl_packed_pixels` _only if_ `gl_glsl_alias_able` is enabled, and otherwise use the old formats, i.e. simply by commenting out the `if (!gl_packed_pixels)` block in gl_vidsdl.c.

**Comment by sezero** [comment]
I just pushed https://github.com/sezero/quakespasm/commit/0c8c70af3c5f0e45935f57b46c3e8aadb82b2d9c to fix the issue only in the GLSL path.

Closing.

**Comment by sezero** [comment]
Pushed https://github.com/sezero/quakespasm/commit/03283182c6c888c5de7d4976709099a9eae7bb33 as a follow-up so users can disable 10 bit depth lightmaps with `-nopackedpixels` if they want to (I experienced noticeable slow-down in some big maps, e.g. ej3_aesop, on old hardware and/or drivers.)
